esp-wifi/src/wifi/mod.rs

```diff
@@ -1383,8 +1388,22 @@ mod asynch {
         ) -> Result<(heapless::Vec<AccessPointInfo, N>, usize), WifiError> {
             Self::clear_events(WifiEvent::ScanDone);
             esp_wifi_result!(wifi_start_scan(false))?;
+
+            struct FreeOnDrop;
+            impl Drop for FreeOnDrop {
+                fn drop(&mut self) {
+                    unsafe {
+                        include::esp_wifi_clear_ap_list();
+                    }
+                }
+            }
+
+            let _guard = FreeOnDrop;
+
             WifiEventFuture::new(WifiEvent::ScanDone).await;
 
+            core::mem::forget(_guard);
```

Comment:
I don't understand what the FreeOnDrop is supposed to do if we immediately forget it

Reply:
It's possible to interrupt the future returned by `scan_n` by putting it in a `select`. In that case, the memory allocated by the wifi stack would not get freed. `FreeOnDrop` is meant to prevent this. It is forgotten right after the future is resolved because the below functions handle freeing memory in their implementations.

Reply:
maybe a comment would be good here?

Reply:
I did that and a bit more: I've renamed it to `FreeApListOnDrop`, and I reused it in the other places where we free on error. Hopefully the result is a bit easier to read, though I'd like to hear from you how it affects understanding the sync code.